include/silo/query_engine/query_result.h

```diff
@@ -17,13 +19,76 @@ struct QueryResultEntry {
+   // Temporary chunk of the query result in the case of streaming, or
+   // the whole result in the case of eager query evaluation.
+   std::vector<QueryResultEntry> query_result_chunk_;
+   // Receives the cleared `query_result_chunk_` and fills in the next
+   // batch of entries of the result set; it adds no entries iff the
+   // end of the result set has been reached.
+   std::function<void(std::vector<QueryResultEntry>&)> get_chunk_;
+   // Iterating through query_result_chunk_.
+   std::vector<QueryResultEntry>::const_iterator iter_;
+   // Used for safety checks when asking to get all values via
+   // `entries` or `entriesMut`.
+   const bool is_materialized_;
+
+  public:
+   /// For eager query evaluation.
+   explicit QueryResult(std::vector<QueryResultEntry>&& query_result)
+       : query_result_chunk_(std::move(query_result)),
+         get_chunk_([](std::vector<QueryResultEntry>& /*query_result_chunk*/) {}),
+         is_materialized_(true) {
+      iter_ = query_result_chunk_.begin();
+   }
+   /// For streaming query evaluation.
+   explicit QueryResult(std::function<bool(std::vector<QueryResultEntry>&)>&& get_chunk)
+       : get_chunk_(get_chunk),
+         is_materialized_(false) {
+      iter_ = query_result_chunk_.begin();
+   }
+   /// The empty result.
+   QueryResult()
+       : get_chunk_([](std::vector<QueryResultEntry>& /*query_result_chunk*/) {}),
+         is_materialized_(true) {}
+
+   // Moves
+   QueryResult(QueryResult&& other) noexcept
+       : query_result_chunk_(std::move(other.query_result_chunk_)),
+         get_chunk_(std::move(other.get_chunk_)),
+         iter_(other.iter_),
+         is_materialized_(other.is_materialized_){};
+   QueryResult& operator=(QueryResult&& other) noexcept {
+      query_result_chunk_ = std::move(other.query_result_chunk_);
+      get_chunk_ = std::move(other.get_chunk_);
+      iter_ = other.iter_;
+      // No marker needed since all fields have one themselves.
+      return *this;
+   }
+   // Copy, only needed for testing::Return in
+   // .conan2/p/b/gtest778a45ec62b47/p/include/gmock/gmock-actions.h
+   QueryResult(QueryResult& other) noexcept = default;
+
+   /// Make the result empty.
+   void clear();
+
+   /// Returns nullptr at the end of the result set. The returned
+   /// pointer is borrowing and is only valid until the next `next`
+   /// call.
+   const QueryResultEntry* next();
+   /// Mutable access to the vector for sorting purposes. Can only be
+   /// called for materialized QueryResult:s. NOTE: this method
+   /// should probably be removed in the future and the code structured
+   /// to never need sorting after creation of the QueryResult.
+   std::vector<QueryResultEntry>& entriesMut();
+   /// Access to the vector for backwards compatibility purposes. Can
+   /// only be called for materialized QueryResult:s. NOTE: this
+   /// method should perhaps be removed in the future.
+   [[nodiscard]] const std::vector<QueryResultEntry>& entries() const;
```

Comment:
Subclasses would admittedly be useful if any code takes them statically, but I'm not sure there is any that will.

Reply:
There has been no case where static type choice would have been appropriate. In fact this is a kind of runtime choice like a variant is. In *either* case, statically those methods have to be available. So whether those methods fail via an exception thrown in a non-shadowed base method, or via the if branch as now, doesn't appear to make a substantial difference to me. But I favour the approach with the boolean because (a) I don't see this as a case of an extensible OO hierarchy and I think it's good to avoid giving the impression, (b) modern practice tends to avoid inheritance (Rust doesn't even offer it), (c) the boolean can use less storage than the pointer to the vtable and the compiled code will be simpler.